### DISCUSSION TITLE
Got 'Type mismatch' error when installed damask-3.0.0 in the step of Cmake compiliation and installation Mehs and Grid Solver

### DISCUSSION THREAD

**Discussion by ruanqc** [damask-multiphysics/DAMASK#38] [Q&A] [answered]
Hi, there. I am trying to insatll the latest version of DAMASK-3.0.0 in Ubuntu 24.04.1 LTS by using the Source Code method. However, I always faild in the step of  cmake "compiliation and installation" of Grid and Mesh Solver. The ERRORs message are as follows, 

$cmake -S damask300 -B build-grid -DDAMASK_SOLVER=grid
$cmake --build build-grid --target install

[  1%] Building Fortran object src/CMakeFiles/DAMASK_grid.dir/parallelization.f90.o
/home/qc/damask/damask300/src/parallelization.f90:201:98:

  201 |   call MPI_Bcast(strlen,1_MPI_INTEGER_KIND,MPI_INTEGER,0_MPI_INTEGER_KIND,MPI_COMM_WORLD, err_MPI)
      |                                                                                                  1
Warning: Procedure ‘mpi_bcast’ called at (1) is not explicitly declared [-Wimplicit-procedure]
/home/qc/damask/damask300/src/parallelization.f90:204:85:

  204 |   call MPI_Bcast(str,strlen,MPI_CHARACTER,0_MPI_INTEGER_KIND,MPI_COMM_WORLD, err_MPI)
      |                                                                                     1
Warning: Procedure ‘mpi_bcast’ called at (1) is not explicitly declared [-Wimplicit-procedure]
/home/qc/damask/damask300/src/parallelization.f90:204:17:

  201 |   call MPI_Bcast(strlen,1_MPI_INTEGER_KIND,MPI_INTEGER,0_MPI_INTEGER_KIND,MPI_COMM_WORLD, err_MPI)
      |                 2
......
  204 |   call MPI_Bcast(str,strlen,MPI_CHARACTER,0_MPI_INTEGER_KIND,MPI_COMM_WORLD, err_MPI)
      |                 1
Error: Type mismatch between actual argument at (1) and actual argument at (2) (CHARACTER(:)/INTEGER(4)).
gmake[2]: *** [src/CMakeFiles/DAMASK_grid.dir/build.make:401: src/CMakeFiles/DAMASK_grid.dir/parallelization.f90.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:98: src/CMakeFiles/DAMASK_grid.dir/all] Error 2
gmake: *** [Makefile:136: all] Error 2
![e1262636e33f65fc7e39b72af7cb5f7](https://github.com/user-attachments/assets/62f87de2-efb3-48b8-8b61-5cd5d7012cb4)

and, also for the mehs solver one, 
$cmake -S damask300 -B build-mesh -DDAMASK_SOLVER=mesh
$cmake --build build-mesh --target install
[  1%] Building Fortran object src/CMakeFiles/DAMASK_mesh.dir/parallelization.f90.o
/home/qc/damask/damask300/src/parallelization.f90:201:98:

  201 |   call MPI_Bcast(strlen,1_MPI_INTEGER_KIND,MPI_INTEGER,0_MPI_INTEGER_KIND,MPI_COMM_WORLD, err_MPI)
      |                                                                                                  1
Warning: Procedure ‘mpi_bcast’ called at (1) is not explicitly declared [-Wimplicit-procedure]
/home/qc/damask/damask300/src/parallelization.f90:204:85:

  204 |   call MPI_Bcast(str,strlen,MPI_CHARACTER,0_MPI_INTEGER_KIND,MPI_COMM_WORLD, err_MPI)
      |                                                                                     1
Warning: Procedure ‘mpi_bcast’ called at (1) is not explicitly declared [-Wimplicit-procedure]
/home/qc/damask/damask300/src/parallelization.f90:204:17:

  201 |   call MPI_Bcast(strlen,1_MPI_INTEGER_KIND,MPI_INTEGER,0_MPI_INTEGER_KIND,MPI_COMM_WORLD, err_MPI)
      |                 2
......
  204 |   call MPI_Bcast(str,strlen,MPI_CHARACTER,0_MPI_INTEGER_KIND,MPI_COMM_WORLD, err_MPI)
      |                 1
Error: Type mismatch between actual argument at (1) and actual argument at (2) (CHARACTER(:)/INTEGER(4)).
gmake[2]: *** [src/CMakeFiles/DAMASK_mesh.dir/build.make:401: src/CMakeFiles/DAMASK_mesh.dir/parallelization.f90.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:98: src/CMakeFiles/DAMASK_mesh.dir/all] Error 2
gmake: *** [Makefile:136: all] Error 2
![image](https://github.com/user-attachments/assets/cc948a12-d35d-4de9-9776-3421197888bc)

The system_report.txt of the PC system is attached following,
[system_report.txt](https://github.com/user-attachments/files/17473800/system_report.txt)

Are there any errors, mistakes, errors, or software version compatibility issues in this installation? I'd very happy If anyone can give me some advices on solving this issure. thank you

**Comment by ruanqc** [comment]
I  found the Q&A (URL: https://www.damask.mpie.de/support/FAQ/installation.html#why-does-compilation-of-grid-or-mesh-solver-fail-due-to-missing-files) there has mentioned this :
"
Why does compilation of grid or mesh solver fail due to an argument mismatch?
This happens when using gfortran 10 and higher together with [MPICH](https://www.mpich.org/). To relax the strict error checking, add -DBUILDCMD_POST='-fallow-argument-mismatch -Wno-pedantic' when invoking CMake.
"

this maily cause cause the problem since the gfortran 13.2.0 is higher then 10 , how could I use this "add -DBUILDCMD_POST='-fallow-argument-mismatch -Wno-pedantic' ", could anyone give some suggestions? thank you

**Comment by ruanqc** [comment]
So everyone, I have finally solve this problem by downgrade  gfortran version gfortran-9.5.

**Reply by cherry12306** [reply]
Hello, after following your advice, I downgraded gfortran to version 9.5. However, I encountered the following issue during computation:[error.md](https://github.com/user-attachments/files/17701925/error.md) I’d like to ask if you've encountered a similar issue before. This seems to be an HDF5 call error.

**Reply by MarDiehl** [reply]
your simulation is not converging and terminates. The HDF5 error is not directly related, it's probably due to different threads/processes not shutting down properly.

**Reply by cherry12306** [reply]
Thank you, Professor MarDiehl. I'll double check my simulation.

**Comment by MarDiehl** [comment, marked as answer]
@ruanqc: good to hear that you solved this. There is another solution, see https://damask-multiphysics.org/support/FAQ/installation.html#why-does-compilation-of-grid-or-mesh-solver-fail-due-to-an-argument-mismatch. I could also imagine that the issue does not exists anymore with newer MPICH version and since your system report shows that you still use PETSc 3.16, I would recommend to update your PETSc installation. Note that DAMASK 3.0.0 does not work with newest PETSc 3.22, but we have a 3.0.1 release available for download that still needs to be announced on the website.

Finally: If you use Ubuntu, the most convenient way to install DAMASK is via the PPA (https://damask-multiphysics.org/installation/package_manager.html#ubuntu). Library versions are not as recent as for a manual installation, but newer than gfortran 9 and PETSc 3.16.

**Reply by ruanqc** [reply]
Thank you Prof. Diehl for this reply. After post this question I have found the solution that as you mentioned why-does-compilation-of-grid-or-mesh-solver-fail-due-to-an-argument-mismatch in the support web page, to be honest I just don't know how to "add -DBUILDCMD_POST='-fallow-argument-mismatch -Wno-pedantic'" and make it work when CMAKE compiles the grid and mehs solver. And finally to use a lower version of the gfortran-9.5. 
I chose to compile and install from source code, mainly because it  might be more easier to implement our particular material model that similar with the existed one, does it can also modify the material model using via PPA installation?   I'd like to try the new version once it is available. Thank you and best regards

**Reply by MarDiehl** [reply]
You have to add the flag when invoking CMake to build the configuration files (the first out of two invocations).
If you want to modify the code, you have indeed to compile from source.

**Reply by ruanqc** [reply]
Thank you professor
